### DISCUSSION TITLE
Add "leave" keyword for escaping exception frames

### DISCUSSION THREAD

**Discussion by asyncawaitmvvm** [dotnet/csharplang#1577] [Language Ideas]
Microsoft's C++ has a __leave keyword for Structured Exception Handling (SEH) which is essentially a "break" for try-catch scopes. It immediately causes flow to exit the try scope and enter the finally scope. It's fantastic for setting up rigid C-style error cleanup without using a goto. This keyword if brought to C# would have strong precedent and get a lot of use.

For reference:
https://msdn.microsoft.com/en-us/library/9xtt5hxz.aspx

> The __leave keyword is valid only within the guarded section of a try-finally statement, and its effect is to jump to the end of the guarded section. Execution continues at the first statement in the termination handler.
> A goto statement can also jump out of the guarded section, but it degrades performance because it invokes stack unwinding. The __leave statement is more efficient because it does not cause stack unwinding.

**Comment by Neme12** [comment]
> It's fantastic for setting up rigid C-style error cleanup without using a goto

Ok... so what is the advantage over goto?

**Comment by stakx** [comment]
> It's fantastic for setting up rigid C-style error cleanup without using a goto. This keyword if brought to C# would have strong precedent and get a lot of use.

"C-style error cleanup" doesn't exactly sound like a tempting thing to have in C#. (Wouldn't error cleanup typically happen inside `catch`, anyways?)

Could you perhaps show an example of what a typical, good use of `leave` might look like?

**Comment by jnm2** [comment]
If I'm understanding right, inverting an `if` in a method results in `if (!condition) return;`. Within a loop it results in `if (!condition) continue;`. With this proposal, if you're within a `try` (or `using`?) block, inverting an if would result in `if (!condition) leave;`?

**Comment by HaloFour** [comment]
> A goto statement can also jump out of the guarded section, but it degrades performance because it invokes stack unwinding. The `__leave` statement is more efficient because it does not cause stack unwinding.

That sounds like a pretty specific optimization for Visual C++.  Considering that C# already emits `leave`/`leave.s` IL opcodes to exit from a protected region is there any reason to believe that a separate `leave` C# keyword would offer any benefits, beyond acting like a `break` statement out of the `try`/`catch` block?
